include/envoy/request_info/dynamic_metadata.h

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/common/pure.h"
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace RequestInfo {
+
+class DynamicMetadata {
+public:
+  virtual ~DynamicMetadata(){};
+
+  // May only be called once for a particular |data_name|
+  template <typename T> void setData(absl::string_view data_name, std::unique_ptr<T>&& data) {
```

Comment:
Please add doxygen-style comments for these methods per the style guide. This method in particular should probably get a note that setData with a duplicate name will destroy data.

Reply:
Done (sorry to have missed this).

In terms of a duplicate name destroying data: If you ask me again I'll do it, but to my mind calling it with a duplicate name is an interface violation, so it feels a little funny to me to say "Don't do this.  Oh, and if you do, the owning pointer you passed in will be destructed, resulting in deallocation of the data you passed."

(It occurs to me that this interface is incomplete, in that it's not possible for a consumer that doesn't have prior knowledge to be sure it isn't going to do something that be an interface contract violation.  I've added a hasData<> method to solve that problem.  Let me know if you want other changes/reversions in this space.)

Reply:
That's fine. I suppose the pointer has to be create inline with the call or else moved, so it shouldn't be surprising that it's freed on error.